File: js/src/action_log.ts
/**
 * Log levels, log-chunk metadata and the process-wide action log sink.
 *
 * The TypeScript twin of `cpp/a11/actions/log.h`. An {@link Action} narrates what
 * it is doing through `Action.log`, which turns the value it is handed into a
 * {@link Chunk} on the reserved {@link ACTION_LOG_OUTPUT} port; this module holds
 * the vocabulary everything that writes or reads such a chunk shares.
 */

import { Chunk } from './data.js';

/** The five severities every language A11 binds to agrees on, quietest first. */
export const LOG_LEVELS = ['debug', 'info', 'warning', 'error', 'critical'] as const;

/** One of {@link LOG_LEVELS}. */
export type LogLevel = (typeof LOG_LEVELS)[number];

/** The level a log written without one carries. */
export const DEFAULT_LOG_LEVEL: LogLevel = 'info';

/**
 * Chunk metadata attribute naming the log's level.
 *
 * The log attributes are unprefixed, unlike {@link CLOSE_STATUS_ATTRIBUTE},
 * because they only ever appear on the reserved log port: there is no user
 * metadata there to collide with.
 */
export const LOG_LEVEL_ATTRIBUTE = 'level';
/** Attribute marking a log as A11's own bookkeeping: `'true'` or `'false'`. */
export const LOG_INTERNAL_ATTRIBUTE = 'internal';
/** Attribute naming a free-form channel a consumer may filter on. */
export const LOG_CHANNEL_ATTRIBUTE = 'channel';
/** Attribute naming the source file the log came from. */
export const LOG_FILE_ATTRIBUTE = 'file';
/** Attribute naming the source line the log came from. */
export const LOG_LINENO_ATTRIBUTE = 'lineno';
/** The value {@link LOG_INTERNAL_ATTRIBUTE} takes when the log is internal. */
export const LOG_INTERNAL_TRUE = 'true';
/** The value it takes when the log is not. */
export const LOG_INTERNAL_FALSE = 'false';

/** Whether `name` is one of {@link LOG_LEVELS}, in either case. */
export function isLogLevel(name: string): boolean {
  return (LOG_LEVELS as readonly string[]).includes(name.toLowerCase());
}

/**
 * The canonical level `name` means, or `null` when it is not one.
 *
 * Accepts `warn` for `warning` and `fatal` for `critical` -- the two spellings
 * host languages differ on. An empty name is {@link DEFAULT_LOG_LEVEL}.
 */
export function parseLogLevel(name: string): LogLevel | null {
  if (name === '') return DEFAULT_LOG_LEVEL;
  const lowered = name.toLowerCase();
  if (lowered === 'warn') return 'warning';
  if (lowered === 'fatal') return 'critical';
  return (LOG_LEVELS as readonly string[]).includes(lowered)
    ? (lowered as LogLevel)
    : null;
}

/** Everything about a log other than the value being logged. */
export interface LogOptions {
  /** Level name; omitted is {@link DEFAULT_LOG_LEVEL}. */
  level?: string;
  /**
   * Media type hint for the serializer.
   *
   * Unlike C++, TypeScript distinguishes text from bytes already: a `string` is
   * `text/plain` and a `Uint8Array` is `application/octet-stream` through the
   * ordinary registry, so a log needs no special case here.
   */
  mimetype?: string;
  /** Logical channel; see {@link LOG_CHANNEL_ATTRIBUTE}. */
  channel?: string;
  /** Source file the log came from. */
  file?: string;
  /** Source line the log came from. */
  lineno?: number;
  /** Whether this is A11's own bookkeeping rather than a user-facing line. */
  internal?: boolean;
  /** Extra attributes, merged *before* the fields above. */
  metadata?: Readonly<Record<string, string | Uint8Array>> | ReadonlyMap<string, string | Uint8Array>;
}

/** One log as a sink sees it. */
export interface LogRecord {
  actionName: string;
  actionId: string;
  level: LogLevel;
  channel: string;
  file: string;
  lineno: number | null;
  internal: boolean;
  mimetype: string;
  data: Uint8Array;
  timestamp: Date | null;
}

/** What the process does with a log it consumes itself. */
export type ActionLogSink = (record: LogRecord) => void;

const decoder = new TextDecoder();

/**
 * Whether a log payload of this media type reads as a line of characters.
 *
 * Text and JSON do, so both print as themselves. Anything else is bytes, and a
 * log line is not the place to render a blob -- {@link logText} describes it
 * instead, and the payload is still on the record for a sink that wants it.
 */
export function isTextualLogMimetype(mimetype: string): boolean {
  if (mimetype.startsWith('text/')) return true;
  const media = mimetype.split(';', 1)[0];
  return media === 'application/json' || media.endsWith('+json');
}

/** A record as one line: its payload where that is text, a description if not. */
export function logText(record: LogRecord): string {
  return isTextualLogMimetype(record.mimetype)
    ? decoder.decode(record.data)
    : `<${record.data.byteLength} bytes of ${record.mimetype}>`;
}

/** The default sink: the console, at the method the level names. */
function reportThroughConsole(record: LogRecord): void {
  const where = record.channel === ''
    ? `[${record.actionName}]`
    : `[${record.actionName}/${record.channel}]`;
  const line = `${where} ${logText(record)}`;
  switch (record.level) {
    case 'debug':
      console.debug(line);
      return;
    case 'warning':
      console.warn(line);
      return;
    case 'error':
    case 'critical':
      console.error(line);
      return;
    default:
      console.info(line);
  }
}

let sink: ActionLogSink | null = null;

/**
 * Install `next` as the process's action log sink.
 *
 * One slot rather than one sink per interested party, so a caller that takes it
 * takes it from whoever had it -- which is what keeps a log from being reported
 * twice. `null` restores the console default.
 */
export function setActionLogSink(next: ActionLogSink | null): void {
  sink = next;
}

/** The installed sink, or the default. Never null. */
export function getActionLogSink(): ActionLogSink {
  return sink ?? reportThroughConsole;
}

/** Report `record` to the installed sink. Never throws. */
export function reportLog(record: LogRecord): void {
  try {
    getActionLogSink()(record);
  } catch {
    // A failure to log must never reach the code that was logging.
  }
}

function attribute(chunk: Chunk, key: string): string {
  const value = chunk.metadata?.attributes.get(key);
  return value === undefined ? '' : decoder.decode(value);
}

/**
 * Read a {@link LogRecord} back out of a log chunk.
 *
 * The inverse of what `Action.log` writes, so a consumer on the far end of a wire
 * reads the metadata the same way every other language does. A missing or
 * unknown level falls back to {@link DEFAULT_LOG_LEVEL}.
 */
export function logRecordFromChunk(
  chunk: Chunk,
  actionName = '',
  actionId = '',
): LogRecord {
  const lineno = Number.parseInt(attribute(chunk, LOG_LINENO_ATTRIBUTE), 10);
  return {
    actionName,
    actionId,
    level: parseLogLevel(attribute(chunk, LOG_LEVEL_ATTRIBUTE)) ?? DEFAULT_LOG_LEVEL,
    channel: attribute(chunk, LOG_CHANNEL_ATTRIBUTE),
    file: attribute(chunk, LOG_FILE_ATTRIBUTE),
    lineno: Number.isNaN(lineno) ? null : lineno,
    internal: attribute(chunk, LOG_INTERNAL_ATTRIBUTE) === LOG_INTERNAL_TRUE,
    mimetype: chunk.mimetype,
    data: chunk.data,
    timestamp: chunk.metadata?.timestamp ?? null,
  };
}

File: kotlin/src/main/kotlin/a11/ActionLog.kt
package a11

/**
 * Log levels, log-chunk metadata and the process-wide action log sink.
 *
 * The Kotlin twin of `cpp/a11/actions/log.h`. An [Action] narrates what it is
 * doing through [Action.log], which turns the value it is handed into a [Chunk]
 * on the reserved [ACTION_LOG_OUTPUT] port; this file holds the vocabulary
 * everything that writes or reads such a chunk shares.
 */

/** The five severities every language A11 binds to agrees on, quietest first. */
val LOG_LEVELS: List<String> = listOf("debug", "info", "warning", "error", "critical")

/** The level a log written without one carries. */
const val DEFAULT_LOG_LEVEL = "info"

/**
 * Chunk metadata attribute naming the log's level.
 *
 * The log attributes are unprefixed, unlike [CLOSE_STATUS_ATTRIBUTE], because
 * they only ever appear on the reserved log port: there is no user metadata there
 * to collide with.
 */
const val LOG_LEVEL_ATTRIBUTE = "level"
/** Attribute marking a log as A11's own bookkeeping: `"true"` or `"false"`. */
const val LOG_INTERNAL_ATTRIBUTE = "internal"
/** Attribute naming a free-form channel a consumer may filter on. */
const val LOG_CHANNEL_ATTRIBUTE = "channel"
/** Attribute naming the source file the log came from. */
const val LOG_FILE_ATTRIBUTE = "file"
/** Attribute naming the source line the log came from. */
const val LOG_LINENO_ATTRIBUTE = "lineno"
/** The value [LOG_INTERNAL_ATTRIBUTE] takes when the log is internal. */
const val LOG_INTERNAL_TRUE = "true"
/** The value it takes when the log is not. */
const val LOG_INTERNAL_FALSE = "false"

/** Whether [name] is one of [LOG_LEVELS], in either case. */
fun isLogLevel(name: String): Boolean = name.lowercase() in LOG_LEVELS

/**
 * The canonical level [name] means, or `null` when it is not one.
 *
 * Accepts `warn` for `warning` and `fatal` for `critical` -- the two spellings
 * host languages differ on. An empty name is [DEFAULT_LOG_LEVEL].
 */
fun parseLogLevel(name: String): String? {
    if (name.isEmpty()) return DEFAULT_LOG_LEVEL
    val lowered = name.lowercase()
    if (lowered == "warn") return "warning"
    if (lowered == "fatal") return "critical"
    return if (lowered in LOG_LEVELS) lowered else null
}

/**
 * Everything about a log other than the value being logged.
 *
 * [metadata] is merged onto the chunk first and the named fields after it, so an
 * explicit [level] wins over a `"level"` the same caller also put in the map.
 */
data class LogOptions(
    val level: String = "",
    /**
     * Media type hint for the serializer.
     *
     * Unlike C++, Kotlin distinguishes text from bytes already: a `String` is
     * `text/plain` and a `ByteArray` is `application/octet-stream` through the
     * ordinary registry, so a log needs no special case here.
     */
    val mimetype: String = "",
    val channel: String = "",
    val file: String = "",
    val lineno: Int? = null,
    val internal: Boolean = false,
    val metadata: Map<String, ByteArray>? = null,
)

/** One log as a sink sees it. */
data class LogRecord(
    val actionName: String,
    val actionId: String,
    val level: String,
    val channel: String,
    val file: String,
    val lineno: Int?,
    val internal: Boolean,
    val mimetype: String,
    val data: ByteArray,
    val timestampMillis: Long?,
)

/** What the process does with a log it consumes itself. */
typealias ActionLogSink = (LogRecord) -> Unit

/**
 * Whether a log payload of this media type reads as a line of characters.
 *
 * Text and JSON do, so both print as themselves. Anything else is bytes, and a
 * log line is not the place to render a blob -- [logText] describes it instead,
 * and the payload is still on the record for a sink that wants it.
 */
fun isTextualLogMimetype(mimetype: String): Boolean {
    if (mimetype.startsWith("text/")) return true
    val media = mimetype.substringBefore(';')
    return media == JSON_MIMETYPE || media.endsWith("+json")
}

/** A record as one line: its payload where that is text, a description if not. */
fun logText(record: LogRecord): String =
    if (isTextualLogMimetype(record.mimetype)) String(record.data, Charsets.UTF_8)
    else "<${record.data.size} bytes of ${record.mimetype}>"

/** The default sink: standard error, with the level and the action in front. */
private fun reportThroughStderr(record: LogRecord) {
    val where =
        if (record.channel.isEmpty()) "[${record.actionName}]"
        else "[${record.actionName}/${record.channel}]"
    System.err.println("${record.level.uppercase()} $where ${logText(record)}")
}

@Volatile private var installedSink: ActionLogSink? = null

/**
 * Install [next] as the process's action log sink.
 *
 * One slot rather than one sink per interested party, so a caller that takes it
 * takes it from whoever had it -- which is what keeps a log from being reported
 * twice. `null` restores the default.
 */
fun setActionLogSink(next: ActionLogSink?) {
    installedSink = next
}

/** The installed sink, or the default. Never null. */
fun getActionLogSink(): ActionLogSink = installedSink ?: ::reportThroughStderr

/** Report [record] to the installed sink. Never throws. */
fun reportLog(record: LogRecord) {
    try {
        getActionLogSink()(record)
    } catch (_: Throwable) {
        // A failure to log must never reach the code that was logging.
    }
}

private fun attribute(chunk: Chunk, key: String): String {
    val value = chunk.metadata?.attributes?.get(key) ?: return ""
    return String(value, Charsets.UTF_8)
}

/**
 * Read a [LogRecord] back out of a log chunk.
 *
 * The inverse of what [Action.log] writes, so a consumer on the far end of a wire
 * reads the metadata the same way every other language does. A missing or unknown
 * level falls back to [DEFAULT_LOG_LEVEL].
 */
fun logRecordFromChunk(chunk: Chunk, actionName: String = "", actionId: String = ""): LogRecord =
    LogRecord(
        actionName = actionName,
        actionId = actionId,
        level = parseLogLevel(attribute(chunk, LOG_LEVEL_ATTRIBUTE)) ?: DEFAULT_LOG_LEVEL,
        channel = attribute(chunk, LOG_CHANNEL_ATTRIBUTE),
        file = attribute(chunk, LOG_FILE_ATTRIBUTE),
        lineno = attribute(chunk, LOG_LINENO_ATTRIBUTE).toIntOrNull(),
        internal = attribute(chunk, LOG_INTERNAL_ATTRIBUTE) == LOG_INTERNAL_TRUE,
        mimetype = chunk.metadata?.mimetype ?: "",
        data = chunk.data,
        timestampMillis = chunk.metadata?.timestampMillis,
    )
